honeycombio/data_source_recipients.go

```diff
@@ -0,0 +1,107 @@
+package honeycombio
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	honeycombio "github.com/honeycombio/terraform-provider-honeycombio/client"
+	"github.com/honeycombio/terraform-provider-honeycombio/honeycombio/internal/hashcode"
+)
+
+func dataSourceHoneycombioRecipients() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceHoneycombioRecipientsRead,
+
+		Description: `
+'honeycombio_recipients' data source provides recipient IDs of recipients matching a set of criteria.
+`,
+
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "The type of recipients.",
+				ValidateFunc: validation.StringInSlice([]string{"email", "pagerduty", "slack", "webhook"}, false),
+			},
+			"detail_filter": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				MinItems:    1,
+				MaxItems:    1,
+				Description: "Attributes to filter the recipients with.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "The name of the detail field to filter by",
+							ValidateFunc: validation.StringInSlice([]string{"address", "channel", "name", "integration_name", "url"}, false),
+						},
+						"value": {
+							Type:     schema.TypeString,
+							Optional: true,
+
+							Description:   "The value of the detail field to match on.",
+							ConflictsWith: []string{"detail_filter.0.value_regex"},
+							ValidateFunc:  validation.NoZeroValues,
+						},
+						"value_regex": {
+							Type:          schema.TypeString,
+							Optional:      true,
+							Description:   "A regular expression string to apply to the value of the detail field to match on.",
+							ConflictsWith: []string{"detail_filter.0.value"},
+							ValidateFunc:  validation.StringIsValidRegExp,
+						},
+					},
+				},
+			},
+			"ids": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
```

Comment:
Is there any reason to restrict this to only returning the IDs instead of having the whole recipient object available?

Reply:
I had initially went about ~50% of the way to return list of Recipient objects and then talked myself out of it looking at [aws_subnets](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnets#attributes-reference).

If you're feeling that makes sense too, I'll do that.

Each object (in JSON for lazy reasons) would be something like:

```json
{
  "id": "jdsfjksdf"
  "type": "pagerduty",
  "integration_key": "ksdfksdfkskdf",
  "integration_name": "my service"
}
```

or

```json
{
  "id": "jdsfj83f"
  "type": "email",
  "address": "alerts@example.net"
}
```

Make sense?

Reply:
I think this means the attributes of the single recipient returned by the `honeycombio_recipient` data source should also populate these for consistency? 🤔

Reply:
AWS data sources are sort of all over the place by the look of it. Eg, [aws_kinesis_stream](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnets#attributes-reference) has quite a few attributes.

I'd say there's no harm in providing more data here, but we may want to skip sensitive information like `integration_key`.

Reply:
Agreed, they are all over the place. But, the "plural" versions tend to just return the IDs (often for use by the singular). For example, you can use `aws_availability_zones` to get all of the AZs matching a filter and then iterate over that list of IDs with `aws_availability_zone`. (there is no `aws_kinesis_streams` for comparison)

... maybe I'll go that route and enrich the single result from `honeycombio_recipient` with all the additional fields? 🤔

Reply:
Ah, interesting! That sounds good 👍

Reply:
Okay, I think we got there.